src/ElectricCar.cpp

```diff
@@ -0,0 +1,35 @@
+#include "ElectricCar.hpp"
+#include "MovingCar.hpp"
+#include "InvalidEngine.hpp"
+
+#include <iostream>
+
+ElectricCar::ElectricCar(ElectricEngine* engine)
+    : electricEngine_(engine) {
+    std::cout << __FUNCTION__ << std::endl;
+}
+
+ElectricCar::~ElectricCar() {
+    std::cout << __FUNCTION__ << std::endl;
+    delete electricEngine_;
+}
+
+void ElectricCar::charge() {
+    std::cout << __FUNCTION__ << std::endl;
+}
+
+void ElectricCar::restore() {
+    std::cout << __FUNCTION__ << std::endl;
+    charge();
+}
+
+void ElectricCar::changeEngine(Engine* engine) {
```

Comment:
Can't you just take ElectricEngine as parameter?

Reply:
In my opinion its good. What if the car is a hybrid? Good thing that we can change different types of engines. The code is more universal. Marcin secured it well. The function throws an exception when the wrong engine type is given.

Reply:
It's an overridden method from Car class `virtual void changeEngine(Engine* engine) = 0;`